File: Projects/Framework-Samples/PLINQO/CSharp/Tracker/Source/Tracker.Tests/DemoTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using CodeSmith.Data.Rules;
using NUnit.Framework;
using Tracker.Core.Data;
using System.Data.Linq;
using CodeSmith.Data.Linq;  

namespace Tracker.Tests
{
    [TestFixture]
    public class DemoTests
    {

        #region Setup
        private const int STATUS_NOT_STARTED = 1;
        private const int STATUS_DONE = 6;
        private const int ROLE_MANAGER = 2;
        private int SpockId
        {
            get;
            set;
        }
        private int JamesId
        { get; set; }

        private int TaskId
        {
            get;
            set;
        }

        [SetUp]
        public void Setup()
        {
            CreateUsers();
            CreateTask(SpockId);
        }

        [TearDown]
        public void TearDown()
        {
            using (var context = new TrackerDataContext())
            {
                context.UserRole.Delete(r => r.UserId == SpockId);
                context.UserRole.Delete(r => r.User.EmailAddress.EndsWith("startrek.com") && r.UserId == SpockId);
                context.Audit.Delete(a => a.TaskId == TaskId || a.UserId == SpockId);
                context.Task.Delete(a => a.CreatedId == SpockId || a.AssignedId == SpockId);
                context.Task.Delete(TaskId);
                context.User.Delete(SpockId);
                context.User.Delete(u => u.EmailAddress.EndsWith("startrek.com"));
            }
        }

        private void CreateTask(int userId)
        {
            var task = new Task()
            {
                AssignedId = userId,
                StatusId = 1,
                Summary = "Explain the visions that inspire you",
                PriorityId = Priority.High,
                CreatedId = userId,
            };
            using (var context = new TrackerDataContext())
            {
                context.Task.InsertOnSubmit(task);
                context.SubmitChanges();
                TaskId = task.Id;
            }
        }

        private void CreateUsers()
        {
            var user = new User()
            {
                FirstName = "Spock",
                LastName = "Sarekson",
                PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA",
                PasswordSalt = "=Unc%",
                EmailAddress = "spock@startrek.com",
                IsApproved = true
            };

            using (var context = new TrackerDataContext())
            {
                context.User.InsertOnSubmit(user);
                context.SubmitChanges();
                SpockId = user.Id;
            }

            var users = new List<User>();

            users.Add(new User()
            {
                FirstName = "James",
                LastName = "Kirk",
                PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA",
                PasswordSalt = "=Unc%",
                EmailAddress = "james.tiberius.kirk@startrek.com",
                IsApproved = true
            });

            users.Add(new User()
            {
                FirstName = "Bones",
                LastName = "McCoy",
                PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA",
                PasswordSalt = "=Unc%",
                EmailAddress = "bones.mccory@startrek.com",
                IsApproved = false
            });

            users.Add(new User()
            {
                FirstName = "Nyota",
                LastName = "Uhura",
                PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA",
                PasswordSalt = "=Unc%",
                EmailAddress = "uhura@startrek.com",
                IsApproved = false
            });

            users.Add(new User()
            {
                FirstName = "Ellen",
                LastName = "Tigh",
                PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA",
                PasswordSalt = "=Unc%",
                EmailAddress = "pavel.chekov@startrek.com",
                IsApproved = false
            });

            using (var context = new TrackerDataContext())
            {
                context.User.InsertAllOnSubmit(users);
                context.SubmitChanges();
                JamesId = context.User.ByFirstName("James").ByLastName("Kirk").First().Id;
            }
        }

        #endregion

        [Test]
        public void Test_Manager_And_Query_Gets()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                Task task = context.Manager.Task.GetByKey(SpockId);
                IQueryable<Task> tasks = context.Manager.Task.GetByAssignedIdStatusId(SpockId, STATUS_NOT_STARTED);
                List<Task> taskList = tasks.ToList();
            }

            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                Task task = context.Task.ByKey(SpockId);
                IQueryable<Task> tasks = context.Task.ByAssignedId(SpockId).ByStatusId(STATUS_NOT_STARTED);
                List<Task> taskList = tasks.ToList();
            }
        }

        [Test]
        public void Test_Many_To_Many()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                User u = context.User.ByKey(SpockId);
                Role r = context.Role.ByName("Manager").First();
                u.RoleList.Add(r);
                context.SubmitChanges();
           }
        }

        [Test]
        public void Test_Enum()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                var task = context.Task.ByKey(TaskId);
                task.PriorityId = Priority.High;
                context.SubmitChanges();
            }
        }

        [Test]
        public void Test_Auditing()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                var user = context.User.ByKey(SpockId);
                user.Comment = "I love my mom, but I hate Winona Ryder.";

                var task = new Task()
                {
                    AssignedId = SpockId,
                    CreatedId = SpockId,
                    StatusId = STATUS_NOT_STARTED,
                    PriorityId = Priority.High,
                    Summary = "Punch Kirk in the face!"
                };
                context.Task.InsertOnSubmit(task);
                context.SubmitChanges();
                Console.Write(context.LastAudit.ToXml());
            }
        }

        [Test]
        public void Test_Rules()
        {
            int brokenRules = 0;
            try
            {
                using (var context = new TrackerDataContext())
                {
                    context.Log = Console.Out;

                    User user = new User();
                    user.EmailAddress = "spock@startrek.com";
                    context.User.InsertOnSubmit(user);
                    context.SubmitChanges();
                }
            }
            catch (BrokenRuleException e)
            {
                brokenRules = e.BrokenRules.Count;
                Console.Write(e.ToString());
            }
            Assert.AreEqual(brokenRules, 5);
        }

        [Test]
        public void Test_Entity_Detach()
        {
            Task task = null;
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                task = context.Task.ByKey(TaskId);
                task.Detach();
            }

            task.StatusId = STATUS_DONE;

            using (var context2 = new TrackerDataContext())
            {
                context2.Log = Console.Out;

                context2.Task.Attach(task, true);
                context2.SubmitChanges();
            }
        }

        [Test]
        public void Test_Entity_Detach_Update()
        {
            Task task = null;
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out; 
                
                task = context.Task.ByKey(TaskId);
                task.Detach();
            }

            using (var context2 = new TrackerDataContext())
            {
                context2.Log = Console.Out;
                // attach, then update properties
                context2.Task.Attach(task);
                task.StatusId = STATUS_DONE;

                context2.SubmitChanges();
            }
        }

        [Test]
        public void Test_Entity_Clone()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                var u = context.Task.ByKey(TaskId);
                Task taskCopy = u.Clone();
                taskCopy.Id = 0;
                context.Task.InsertOnSubmit(taskCopy);
                context.SubmitChanges();
            }
        }

        [Test]
        public void Test_Serialization()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                // Write to Console/Output ... or break point.

                var task = context.Task.ByKey(TaskId);
                var s = task.ToXml();
                Console.Write(s);
            }
        }

        [Test]
        public void Test_Query_Result_Cache()
        {
            using (var context = new TrackerDataContext())
            {
                //By default, the cached results use a one minute sliding expiration with
                //no absolute expiration.
                var tasks = context.Task.ByAssignedId(SpockId).FromCache();
                var cachedTasks = context.Task.ByAssignedId(SpockId).FromCache();

                //query result is now cached 300 seconds
                var approvedUsers = context.User.ByIsApproved(true).FromCache(300);
                var cachedApprovedUsers = context.User.ByIsApproved(true).FromCache(300);
            }
        }

        [Test]
        public void Test_Batch_Update()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;
                
                context.Task.Update(
                    u => u.StatusId == STATUS_NOT_STARTED, 
                    u2 => new Task() { StatusId = STATUS_DONE });

                var tasks = from t in context.Task
                            where t.StatusId == STATUS_DONE
                            select t;
                context.Task.Update(tasks, u => new Task { StatusId = STATUS_NOT_STARTED });

            }
        }

        [Test]
        public void Test_Batch_Delete()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                
                context.User.Delete(JamesId);

                context.User.Delete(u => u.IsApproved == false && u.LastName != "McCoy" && u.EmailAddress.EndsWith("startrek.com"));

                IQueryable<User> usersToDelete = from u in context.User
                                                 where u.IsApproved == false && u.LastName == "McCoy" 
                                                 select u;
                context.User.Delete(usersToDelete);
            }
        }

        [Test]
        public void Test_Stored_Procedure_with_Multiple_Results()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                // Create Procedure [dbo].[GetUsersWithRoles]
                // As
                // Select * From [User]
                // Select * From UserRole
                // GO

                var results = context.GetUsersWithRoles();
                List<User> users = results.GetResult<User>().ToList();
                List<UserRole> roles = results.GetResult<UserRole>().ToList();
            }
        }

        [Test]
        public void Test_Batch_Queries()
        {
            using (var context = new TrackerDataContext())
            {
                context.Log = Console.Out;

                var q1 = from u in context.User select u;
                var q2 = from ur in context.UserRole select ur;
                IMultipleResults results = context.ExecuteQuery(q1, q2);
                List<User> users = results.GetResult<User>().ToList();
                List<UserRole> roles = results.GetResult<UserRole>().ToList();
            }
        }

        
    }
}

File: trunk/Projects/Framework-Samples/PLINQO/VisualBasic/Tracker/Source/Tracker.Tests/DemoTests.vb
Imports System.Collections.Generic
Imports System.Data.Linq
Imports System.Linq
Imports CodeSmith.Data.Linq
Imports CodeSmith.Data.Rules
Imports NUnit.Framework
Imports Tracker.Core.Data

Namespace Tracker.Tests


    <TestFixture()> _
    Public Class DemoTests
#Region "Setup"

        Private Const STATUS_NOT_STARTED As Integer = 1
        Private Const STATUS_DONE As Integer = 6
        Private Const ROLE_MANAGER As Integer = 2
        Private Property SpockId() As Integer
            Get
                Return m_SpockId
            End Get
            Set(ByVal value As Integer)
                m_SpockId = value
            End Set
        End Property
        Private m_SpockId As Integer
        Private Property JamesId() As Integer
            Get
                Return m_JamesId
            End Get
            Set(ByVal value As Integer)
                m_JamesId = value
            End Set
        End Property
        Private m_JamesId As Integer

        Private Property TaskId() As Integer
            Get
                Return m_TaskId
            End Get
            Set(ByVal value As Integer)
                m_TaskId = value
            End Set
        End Property
        Private m_TaskId As Integer

        <TestFixtureSetUp()> _
        Public Sub Setup()
            TearDown()
            CreateUsers()
            CreateTask(SpockId)
        End Sub

        <TestFixtureTearDown()> _
        Public Sub TearDown()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                db.Audit.Delete(Function(a) a.User.EmailAddress.EndsWith("startrek.com"))
                db.UserRole.Delete(Function(ur) ur.User.EmailAddress.EndsWith("startrek.com"))
                db.Task.Delete(Function(t) t.AssignedUser.EmailAddress.EndsWith("startrek.com") OrElse t.CreatedUser.EmailAddress.EndsWith("startrek.com"))
                db.User.Delete(Function(u) u.EmailAddress.EndsWith("startrek.com"))
            End Using
        End Sub

        Private Sub CreateTask(ByVal userId As Integer)
            Dim task As New Task() With { _
             .AssignedId = userId, _
             .Status = Status.InProgress, _
             .Summary = "Explain the visions that inspire you", _
             .Priority = Priority.High, _
             .CreatedId = userId _
            }
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                db.Task.InsertOnSubmit(task)
                db.SubmitChanges()
                TaskId = task.Id
            End Using
        End Sub

        Private Sub CreateUsers()
            Dim user As New User() With { _
             .FirstName = "Spock", _
             .LastName = "Sarekson", _
             .PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA", _
             .PasswordSalt = "=Unc%", _
             .EmailAddress = "spock@startrek.com", _
             .IsApproved = True _
            }

            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                db.User.InsertOnSubmit(user)
                db.SubmitChanges()
                SpockId = user.Id
            End Using

            Dim users As New List(Of User)()

            users.Add(New User() With { _
             .FirstName = "James", _
             .LastName = "Kirk", _
             .PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA", _
             .PasswordSalt = "=Unc%", _
             .EmailAddress = "james.tiberius.kirk@startrek.com", _
             .IsApproved = True _
            })

            users.Add(New User() With { _
             .FirstName = "Bones", _
             .LastName = "McCoy", _
             .PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA", _
             .PasswordSalt = "=Unc%", _
             .EmailAddress = "bones.mccory@startrek.com", _
             .IsApproved = False _
            })

            users.Add(New User() With { _
             .FirstName = "Nyota", _
             .LastName = "Uhura", _
             .PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA", _
             .PasswordSalt = "=Unc%", _
             .EmailAddress = "uhura@startrek.com", _
             .IsApproved = False _
            })

            users.Add(New User() With { _
             .FirstName = "Ellen", _
             .LastName = "Tigh", _
             .PasswordHash = "aM/Vndh7cYd3Mxq7msArjl9YU8zoR6fF+sVTSUCcsJi2bx+cwOI0/Bkr5hfq9vYfTe3/rlgPpSMg108acpw+qA", _
             .PasswordSalt = "=Unc%", _
             .EmailAddress = "ellen.tigh@startrek.com", _
             .IsApproved = False _
            })

            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                db.User.InsertAllOnSubmit(users)
                db.SubmitChanges()
                JamesId = db.User.ByFirstName("James").ByLastName("Kirk").First().Id
            End Using
        End Sub

#End Region

        <Test()> _
        Public Sub Test_Manager_And_Query_Gets()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim task As Task = db.Manager.Task.GetByKey(SpockId)
                Dim tasks As IQueryable(Of Task) = db.Manager.Task.GetByAssignedIdStatus(SpockId, Status.InProgress)
                Dim taskList As List(Of Task) = tasks.ToList()
            End Using

            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim task As Task = db.Task.GetByKey(SpockId)
                Dim tasks As IQueryable(Of Task) = db.Task.ByAssignedId(SpockId).ByStatus(Status.InProgress)
                Dim taskList As List(Of Task) = tasks.ToList()
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Query_Advanced()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim priorites As List(Of Task) = db.Task.Where(Function(t) Not t.Priority.HasValue OrElse t.Priority.Value = Priority.High).ToList()
                Dim something As List(Of Task) = db.Task.ByPriority(Nothing, Priority.High).ToList()
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Many_To_Many()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim u As User = db.User.GetByKey(SpockId)
                Dim r As Role = db.Role.ByName("Manager").First()
                u.RoleList.Add(r)
                db.SubmitChanges()
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Enum()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim task As Task = db.Task.GetByKey(TaskId)
                task.Priority = Priority.High
                db.SubmitChanges()
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Auditing()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim user As User = db.User.GetByKey(SpockId)
                user.Comment = "I love my mom, but I hate Winona Ryder."

                Dim task As New Task() With { _
                 .AssignedId = SpockId, _
                 .CreatedId = SpockId, _
                 .Status = Status.NotStarted, _
                 .Priority = Priority.High, _
                 .Summary = "Punch Kirk in the face!" _
                }
                db.Task.InsertOnSubmit(task)
                db.SubmitChanges()
                Console.Write(db.LastAudit.ToXml())
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Rules()
            Dim brokenRules As Integer = 0
            Try
                Using db As New TrackerDataContext() With { _
                 .Log = Console.Out _
                }
                    Dim user As New User()
                    user.EmailAddress = "spock@startrek.com"
                    db.User.InsertOnSubmit(user)
                    db.SubmitChanges()
                End Using
            Catch e As BrokenRuleException
                brokenRules = e.BrokenRules.Count
                Console.Write(e.ToString())
            End Try
            Assert.AreEqual(brokenRules, 2)
        End Sub

        <Test()> _
        Public Sub Test_Entity_Detach()
            Dim task As Task = Nothing
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                task = db.Task.GetByKey(TaskId)
                task.Detach()
            End Using

            task.Status = Status.Done

            Using context2 As New TrackerDataContext()
                context2.Log = Console.Out

                context2.Task.Attach(task, True)
                context2.SubmitChanges()
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Entity_Detach_Update()
            Dim task As Task = Nothing
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                task = db.Task.GetByKey(TaskId)
                task.Detach()
            End Using

            Using context2 As New TrackerDataContext()
                context2.Log = Console.Out
                ' attach, then update properties
                context2.Task.Attach(task)
                task.Status = Status.Done

                context2.SubmitChanges()
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Entity_Clone()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim u As Task = db.Task.GetByKey(TaskId)
                Dim taskCopy As Task = u.Clone()
                taskCopy.Id = 0
                db.Task.InsertOnSubmit(taskCopy)
                db.SubmitChanges()
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Serialization()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                ' Write to Console/Output ... or break point.

                Dim task As Task = db.Task.GetByKey(TaskId)
                Dim s As String = task.ToXml()
                Console.Write(s)
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Query_Result_Cache()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                'By default, the cached results use a one minute sliding expiration with
                'no absolute expiration.
                Dim tasks As IEnumerable(Of Task) = db.Task.ByAssignedId(SpockId).FromCache()
                Dim cachedTasks As IEnumerable(Of Task) = db.Task.ByAssignedId(SpockId).FromCache()

                'query result is now cached 300 seconds
                Dim approvedUsers As IEnumerable(Of User) = db.User.ByIsApproved(True).FromCache(300)
                Dim cachedApprovedUsers As IEnumerable(Of User) = db.User.ByIsApproved(True).FromCache(300)
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Batch_Update()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                db.Task.Update(Function(u) u.Status = Status.NotStarted, Function(u2) New Task() With { _
        .Status = Status.Done _
       })

                Dim tasks As IQueryable(Of Task) = From t In db.Task _
                 Where t.Status = Status.Done _
                 Select t
                db.Task.Update(tasks, Function(u) New Task() With { _
        .Status = Status.NotStarted _
       })
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Batch_Delete()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                db.User.Delete(JamesId)

                db.User.Delete(Function(u) u.IsApproved = False AndAlso u.LastName <> "McCoy" AndAlso u.EmailAddress.EndsWith("startrek.com"))

                Dim usersToDelete As IQueryable(Of User) = From u In db.User _
                 Where u.IsApproved = False AndAlso u.LastName = "McCoy" _
                 Select u
                db.User.Delete(usersToDelete)
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Stored_Procedure_with_Multiple_Results()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                ' Create Procedure [dbo].[GetUsersWithRoles]
                ' As
                ' Select * From [User]
                ' Select * From UserRole
                ' GO

                Dim results As IMultipleResults = db.GetUsersWithRoles()
                Dim users As List(Of User) = results.GetResult(Of User)().ToList()
                Dim roles As List(Of UserRole) = results.GetResult(Of UserRole)().ToList()
            End Using
        End Sub

        <Test()> _
        Public Sub Test_Batch_Queries()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim q1 As IQueryable(Of User) = From u In db.User _
                 Select u
                Dim q2 As IQueryable(Of UserRole) = From ur In db.UserRole _
                 Select ur
                Dim results As IMultipleResults = db.ExecuteQuery(q1, q2)
                Dim users As List(Of User) = results.GetResult(Of User)().ToList()
                Dim roles As List(Of UserRole) = results.GetResult(Of UserRole)().ToList()
            End Using
        End Sub

        <Test()> _
        Public Sub Test_ToPagedList()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim q1 As IOrderedQueryable(Of User) = From u In db.User _
                 Order By u.EmailAddress _
                 Select u

                Dim users As PagedList(Of User) = q1.ToPagedList(0, 5)

                Assert.IsNotNull(users)
                Assert.AreEqual(5, users.Count)
            End Using
        End Sub
    End Class
End Namespace
